### PR TITLE
Withdraw indexing rewards to beneficiary address with thawing period

### PR DESCRIPTION
**Pull request by abarmat**
This PR introduces a number of changes in how rewards are managed:

### Main Changes

- Indexers can set whether to stake rewards or send them to a rewards pool by passing a __restake_ boolean to both closeAllocation() and claim(). This allows efficient and flexible change of strategy from the indexer agent.

- A function called `withdrawRewards(address _beneficiary)` is available to claim accrued tokens, according to a thawing period, from both indexing rewards and query fees that were send to the rewards pool. This function allows withdrawing rewards to an arbitrary address.

### Additional Considerations

- Perform some changes to reduce the bytecode size of Staking contract, such as moving some math logic to `MathUtils`, use `_pushTokens()` and `_pullTokens()` for token operations.
- Move structs that are related to Staking contract storage to IStakingData file.

Related-to: GIP-0004
Discussed in: https://forum.thegraph.com/t/rewards-destination-for-indexers-indexing-rewards/1279/64

### CODE REVIEW

File: test/rewards/rewards.test.ts [hunk 1]
```diff
@@ -594,13 +601,69 @@ describe('Rewards', () => {
...
+        // NOTE: calculated manually on a spreadsheet
+        const expectedIndexerStake = beforeIndexer1Stake
+        const expectedTokenSupply = beforeTokenSupply.add(expectedIndexingRewards)
+        // Check stake should not have changed
+        expect(toRound(afterIndexer1Stake)).eq(toRound(expectedIndexerStake))
+        // Check indexer balance remains the same
+        expect(afterIndexer1Balance).eq(beforeIndexer1Balance)
+        // Check indexing rewards are kept in the staking contract
+        expect(toRound(afterStakingBalance)).eq(
+          toRound(beforeStakingBalance.add(expectedIndexingRewards)),
```
Comment (davekaj) [inline test/rewards/rewards.test.ts:665]: No checking of the rewards pool in this test

File: contracts/staking/Staking.sol [hunk 2]
```diff
@@ -1563,9 +1592,29 @@ contract Staking is StakingV1Storage, GraphUpgradeable, IStaking {
...
+        bool _restake
+    ) private {
+        if (_amount == 0) return;
+
+        if (_restake) {
+            // Restake to place fees into the indexer stake
+            _stake(_beneficiary, _amount);
+        } else {
+            // Transfer funds to the beneficiary's rewards pool
+            rewardsPools[_beneficiary].lockTokens(_amount, thawingPeriod);
```
Comment (davekaj) [inline contracts/staking/Staking.sol:1617]: i could see adding an event here being useful

**Review by davekaj — CHANGES_REQUESTED** [review]

File: contracts/staking/Staking.sol [hunk 3]
```diff
@@ -518,8 +525,8 @@ contract Staking is StakingV1Storage, GraphUpgradeable, IStaking {
      * @param _indexer Address of the indexer
      * @return True if indexer has staked tokens
      */
-    function hasStake(address _indexer) external override view returns (bool) {
-        return stakes[_indexer].hasTokens();
+    function hasStake(address _indexer) external view override returns (bool) {
+        return stakes[_indexer].tokensStaked > 0;
```
[the remark this answers is not in the document]
Comment (abarmat) [reply contracts/staking/Staking.sol:529]: The reason I did that is effectively reduced contract bytecode size

File: contracts/staking/libs/Stakes.sol [hunk 4]
```diff
@@ -71,7 +73,12 @@ library Stakes {
         // Take into account period averaging for multiple unstake requests
         uint256 lockingPeriod = _period;
         if (stake.tokensLocked > 0) {
-            lockingPeriod = stake.getLockingPeriod(_tokens, _period);
+            lockingPeriod = MathUtils.weightedAverage(
```
[the remark this answers is not in the document]
Comment (abarmat) [reply contracts/staking/libs/Stakes.sol:76]: We can do it for all the parameters that access the library struct in this file, what do you think?

File: contracts/staking/Staking.sol [hunk 5]
```diff
@@ -717,6 +722,24 @@ contract Staking is StakingV1Storage, GraphUpgradeable, IStaking {
...
 
+    /**
+     * @dev Withdraw all available rewards.
+     * @param _beneficiary Address to send rewards
+     */
+    function withdrawRewards(address _beneficiary) external override notPaused {
+        Rewards.Pool memory rewardsPool = rewardsPools[msg.sender];
+
+        require(rewardsPool.tokensLocked > 0, "rewards-empty");
+        require(rewardsPool.tokensLockedUntil >= block.number, "rewards-locked");
```
[the remark this answers is not in the document]
Comment (abarmat) [reply contracts/staking/Staking.sol:733]: Good catch

File: contracts/staking/Staking.sol [hunk 6]
```diff
@@ -1563,9 +1592,29 @@ contract Staking is StakingV1Storage, GraphUpgradeable, IStaking {
...
-        if (indexerRewards > 0) {
-            _stake(_indexer, indexerRewards);
+        // Send the indexing rewards to the indexer
+        _sendRewards(indexerRewards, _indexer, _restake);
+    }
+
+    /**
+     * @dev Send rewards to the appropiate destination.
+     * @param _amount Number of rewards tokens
+     * @param _beneficiary Address of the beneficiary of rewards
```
[the remark this answers is not in the document]
Comment (abarmat) [reply contracts/staking/Staking.sol:1602]: I think it's a good idea

File: test/rewards/rewards.test.ts (hunk 1 above)
Comment (abarmat) [reply test/rewards/rewards.test.ts:665]: +1

File: contracts/staking/Staking.sol (hunk 2 above)
Comment (abarmat) [reply contracts/staking/Staking.sol:1617]: +1

File: contracts/staking/Staking.sol [hunk 7]
```diff
@@ -717,6 +722,24 @@ contract Staking is StakingV1Storage, GraphUpgradeable, IStaking {
         _withdraw(msg.sender);
     }
 
+    /**
+     * @dev Withdraw all available rewards.
+     * @param _beneficiary Address to send rewards
+     */
+    function withdrawRewards(address _beneficiary) external override notPaused {
```
[the remark this answers is not in the document]
Comment (abarmat) [reply contracts/staking/Staking.sol:729]: +1

File: contracts/staking/libs/Stakes.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (davekaj) [reply contracts/staking/libs/Stakes.sol:76]: i think it is a good pattern lets do it

**Comment by abarmat** [comment]
Archive as alternative research solution

**Pull request closed without merging**